Client script test runner

Polyrhythm opened the pull request:
**Impetus**:
Current client-side script testing is basically non-existant. There are a handful of test scripts in `scripts/developer/tests` that don't really conform to any given methodology and are not being run as part of a regular QA testing regimen. Ensuring that these script tests don't break is just as important as *.cpp tests and should therefore be part of the QA process for every PR.

**Solution**:
Rely on Jasmine to run all scripting tests. For convenience, I have built a developer menu item that kicks off a `JSConsole` which runs an included test runner that, in turn, includes all unit tests and runs them.

**Features**:
- [x] [Custom console output](https://github.com/highfidelity/hifi/compare/master...Polyrhythm:ryan/client-side-tests?expand=1#diff-aa4bdc06d9d7e49754a280641554803d) clearly showing what passes, what fails (with complete stack trace), how long each individual spec takes, and how long all specs take in aggregate
- [x] Menu item for running and viewing test results for QA convenience - one click process - very clear if any tests fail
- [x] [Consolidated test runner](https://github.com/highfidelity/hifi/compare/master...Polyrhythm:ryan/client-side-tests?expand=1#diff-d406174eb48a210897d225c829bed016) where all unit tests should be added
- [x] [Improved bind unit test](https://github.com/highfidelity/hifi/compare/master...Polyrhythm:ryan/client-side-tests?expand=1#diff-59adb16052dd9c1be6fee63be11c5204) and example Entity unit test for showing the general use case

**Known issues**:
Asynchronous testing is broken right now - Jasmine supplies a `done()` callback for making this work which is not being called in unit tests right now. I found this out while attempting to run an existing unit test in the repo which was broken at some point between when it was first introduced and now - a good case for having regular client-side testing as part of QA.

**QA**:
1. Connect to a domain with edit privileges (tests will involve the editing tools)
1. Developer -> Tests -> Run Client Script Tests`

Passing specs:
![pass](https://cloud.githubusercontent.com/assets/1242927/20874282/377249d2-ba65-11e6-9f88-b04fa44add7f.png)

Failing spec with error output:
![fail](https://cloud.githubusercontent.com/assets/1242927/20874283/3a71cf54-ba65-11e6-8426-517cd4458e67.png)

**Devs**:
If you want to mess around with this and try to introduce test breakage on purpose, you can either edit the test scripts yourself in the main repo and then build to propagate changes to the build-exported scripts, or you can just go ahead and manually edit the scripts that are pulled in to whichever build dir you happen to be using for interface.

hifi-gustavo commented:
Build console output for commit: 39e1a3fb4444d5cabc4d077bef8d39daf627dd42 available at: https://gist.github.com/4563b6e5d18793ddc2113fc0f5f3d337<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9148-39e1a3f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9148-39e1a3f.dmg

cozza13 commented:
This looks great. I will get this tested.

OmegaHeron commented:
Testing Windows 10
QA Pass
General smoke test pass.
Running new test menu option I get expected console output.

huffman commented on interface/src/Menu.cpp:
```diff
@@ -589,7 +589,9 @@ Menu::Menu() {
-
+	// Developer >> Tests >>>
+	MenuWrapper* testMenu = developerMenu->addMenu("Tests");
+	addActionToQMenuAndActionHash(testMenu, MenuOption::RunClientScriptTests, 0, dialogsManager.data(), SLOT(showTestingResults()));
```
Indentation should be 4 spaces (no tabs)

Polyrhythm replied:
👍 I have to get my IDE to respect the `.editorconfig` to stop this nonsense :)

huffman commented on interface/src/ui/TestingDialog.cpp:
```diff
@@ -0,0 +1,39 @@
+	setAttribute(Qt::WA_DeleteOnClose);
+	setWindowTitle(windowLabel);
+
+	_console = new JSConsole(this);
```
Not a blocker, but unique_ptr+shared_ptr are typically preferred over bare pointers.

huffman commented on interface/src/ui/TestingDialog.cpp:
```diff
@@ -0,0 +1,39 @@
+
+void TestingDialog::onTestingFinished(const QString& scriptPath) {
+	_engine = NULL;
+	_console->setScriptEngine(NULL);
```
Coding standard: use `nullptr` not `NULL` (Is this in the docs? Maybe that needs to be added)

Polyrhythm replied:
Will fix and also hit some other places that I think uses this exact line where I inferred the style.

huffman commented on scripts/developer/tests/unit_tests/avatarUnitTests.js:
```diff
@@ -17,6 +14,7 @@ describe("MyAvatar", function () {
 
         // wait until we are finished loading
         var id = Script.setInterval(function () {
+            print(MyAvatar.jointNames.length);
```
Was this meant to be left here?

Polyrhythm replied:
That `print` statement can be taken out of the diff. I will fix!

huffman commented:
Great stuff, I have something I was writing tests for that this will be useful for.

Polyrhythm commented:
* converted all tabs to spaces (btw: used a [visual studio plugin](https://marketplace.visualstudio.com/items?itemName=EditorConfigTeam.EditorConfig) to respect the `.editorconfig` for this, recommended
* changed the raw ptr to a unique_ptr and got rid of the now-redundant destructor for `TestingDialog.cpp`
* removed `print` diff on AvatarUnitTest

hifi-gustavo commented:
Build console output for commit: 094857949d9d978a8916282c1c406c33991701f4 available at: https://gist.github.com/4563b6e5d18793ddc2113fc0f5f3d337<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9148-0948579.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9148-0948579.dmg
